### PR TITLE
Allow CPU and CUDA kernels to wait on each other

### PR DESCRIPTION
**Pull request by lcw**
This allows the event system to be used to coordinate work done on the
host with work done on the device.

### CODE REVIEW

**Comment by lcw** [comment]
bors try

File: src/backends/cpu.jl [hunk 1]
```diff
@@ -41,7 +41,15 @@ end
 function __run(obj, ndrange, iterspace, args, dependencies)
     return Threads.@spawn begin
         if dependencies !== nothing
-            Base.sync_end(map(e->e.task, dependencies))
+            cpu_tasks = Core.Task[]
+            for event in dependencies
+                if event isa CPUEvent
+                    push!(cpu_tasks, event.task)
+                else
+                    wait(event)
```
Comment (vchuravy) [inline src/backends/cpu.jl]:
```suggestion
                    wait(event, ()->yield())
```

I am worried about busy-waiting on the GPU causing us the block a thread from making progress. So we should pass a CPU yield as remedy.

File: src/backends/cpu.jl [hunk 2]
```diff
@@ -41,7 +41,15 @@ end
 function __run(obj, ndrange, iterspace, args, dependencies)
     return Threads.@spawn begin
         if dependencies !== nothing
-            Base.sync_end(map(e->e.task, dependencies))
+            cpu_tasks = Core.Task[]
+            for event in dependencies
```
Comment (vchuravy) [inline src/backends/cpu.jl]: In general I would invert this scheme to first wait on a CPU events, and then wait on the GPU events.
The CPU wait is more efficiently implemented in the scheduler and we don't have to busy-wait on the GPU.

File: src/backends/cpu.jl (hunk 1 above)
Comment (lcw) [reply src/backends/cpu.jl]: Done, and added a yield in the cuda backend as well.

**Comment by lcw** [comment]
bors try

File: src/backends/cpu.jl (hunk 2 above)
Comment (lcw) [reply src/backends/cpu.jl]: Okay, I did this and rearranged the cuda backed to first enqueue the gpu waits and then wait on the CPU.

**Comment by vchuravy** [comment]
bors try

**Pull request merged**